exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/common/HashTableTemplate.java

```diff
@@ -224,6 +224,13 @@ private boolean isKeyMatch(int incomingRowIdx,
       if (isProbe) {
         match = isKeyMatchInternalProbe(incomingRowIdx, currentIdxWithinBatch);
       } else {
+        // in case (of a hash-join only) where both the new incoming key and the current are null, treat them as
+        // a match; i.e. the new would be added into the helper (but not the Hash-Table !), though it would never
+        // be used (not putting it into the helper would take a bigger code change, and some performance cost, hence
+        // not worth it).  In the past such a new null key was added into the Hash-Table (i.e., no match), which
+        // created long costly chains - SEE DRILL-6880)
+        if ( areBothKeysNull(incomingRowIdx, currentIdxWithinBatch) ) { return true; }
```

Comment:
The code on old line 230 is in a way "not clean", as it violates the method's name/signature (which implies just a functional comparison of two keys). That code advances to the next key in the hash-table chain in case of a no-match. The "clean way" would be to make this code a separate "advanceToNextIdx" method, which actually fits well in the for() loop where _isKeyMatch()_ is used.

Reply:
Agree that the function name does not reflect the fact that the index gets incremented.  Either we change the function name to `isKeyMatchAndUpdateIndex()` or do that adjustment outside this function.